File: Packages/TeamTasksKit/Sources/TeamTasksCore/Errors/AppError.swift
import Foundation

/// Every error surfaced to the UI. Backends map their errors to these cases (see `BackendErrorMapper`).
public enum AppError: Error, Sendable, Hashable {
    // Auth
    case notAuthenticated
    case invalidCredentials
    case emailAlreadyUsed
    case weakPassword
    case invalidEmail
    case otpInvalid
    case emailRateLimited
    case emailNotConfirmed

    // Input validation
    case invalidDisplayName
    case invalidName
    case invalidTitle
    case invalidDetails
    case invalidInput

    // Groups / membership
    case invalidCode
    case rateLimited
    case lastAdmin
    case notMember
    case cannotRemoveSelf

    // Tasks
    case assigneeNotMember
    case tooManyAssignees

    // v2 (docs/CONTRACTS-V2.md §3)
    /// `invalid_color`, `invalid_emoji`: a group or avatar color / emoji.
    case invalidAppearance
    /// `invalid_recurrence`: the rule's shape (interval, weekdays, time zone).
    case invalidRecurrence
    /// `recurrence_requires_due_date`: a recurring task without due date.
    case recurrenceNeedsDueDate
    /// `invalid_rotation`: 2–20 distinct members, on a recurring task (also `set_task_assignees` on a rotating task).
    case invalidRotation
    /// `invalid_item_title`: a checklist item title.
    case invalidChecklistItem
    /// `too_many_items`: more than 30 checklist items.
    case tooManyChecklistItems

    // Generic
    case forbidden
    case forbiddenFields
    case notFound
    case conflict
    case network
    case misconfigured
    case unknown(String)

    /// Message shown to the user (French UI).
    public var messageFR: String {
        switch self {
        case .notAuthenticated: "Votre session a expiré. Reconnectez-vous."
        case .invalidCredentials: "E-mail ou mot de passe incorrect."
        case .emailAlreadyUsed: "Un compte existe déjà avec cet e-mail."
        case .weakPassword: "Mot de passe trop faible (8 caractères minimum)."
        case .invalidEmail: "Adresse e-mail invalide."
        case .otpInvalid: "Code invalide ou expiré."
        case .emailRateLimited: "Trop d’e-mails envoyés. Réessayez plus tard."
        case .emailNotConfirmed: "Confirmez d’abord votre adresse e-mail."
        case .invalidDisplayName: "Le nom doit contenir entre 1 et 50 caractères."
        case .invalidName: "Le nom du groupe doit contenir entre 1 et 60 caractères."
        case .invalidTitle: "Le titre doit contenir entre 1 et 200 caractères."
        case .invalidDetails: "La description ne doit pas dépasser 5000 caractères."
        case .invalidInput: "Certaines informations sont invalides."
        case .invalidCode: "Code d’invitation invalide."
        case .rateLimited: "Trop de tentatives. Réessayez dans une heure."
        case .lastAdmin: "Vous êtes le seul admin\u{00A0}: nommez d’abord un autre admin."
        case .notMember: "Cette personne ne fait pas partie du groupe."
        case .cannotRemoveSelf: "Utilisez «\u{00A0}Quitter le groupe\u{00A0}» pour vous retirer vous-même."
        case .assigneeNotMember: "Une personne assignée ne fait pas partie du groupe."
        case .tooManyAssignees: "20 personnes assignées au maximum."
        case .invalidAppearance: "Couleur ou emoji invalide."
        case .invalidRecurrence: "Répétition invalide."
        case .recurrenceNeedsDueDate: "Choisissez une échéance pour répéter la tâche."
        case .invalidRotation: "Le tour de rôle demande de 2 à 20 membres du groupe."
        case .invalidChecklistItem: "Un élément doit contenir entre 1 et 200 caractères."
        case .tooManyChecklistItems: "30 éléments au maximum."
        case .forbidden: "Action non autorisée."
        case .forbiddenFields: "Vous pouvez seulement changer le statut de cette tâche."
        case .notFound: "Élément introuvable. Il a peut-être été supprimé."
        case .conflict: "Cet élément existe déjà."
        case .network: "Connexion impossible. Vérifiez votre réseau."
        case .misconfigured: "L’application n’est pas configurée (Supabase)."
        case let .unknown(detail): "Une erreur est survenue. (\(detail))"
        }
    }
}

extension AppError: LocalizedError {
    public var errorDescription: String? { messageFR }
}

extension AppError {
    /// Wraps any error into an `AppError` (already-mapped errors pass through).
    public static func wrap(_ error: any Error) -> AppError {
        if let appError = error as? AppError { return appError }
        if error is CancellationError { return .unknown("annulé") }
        return .unknown(String(describing: error))
    }
}

File: mobile/src/core/appError.ts
/** Every error kind surfaced to the UI (Swift `AppError`); backends map their errors to these kinds. */
export const APP_ERROR_KINDS = [
  // Auth
  'notAuthenticated',
  'invalidCredentials',
  'emailAlreadyUsed',
  'weakPassword',
  'invalidEmail',
  'otpInvalid',
  'emailRateLimited',
  'emailNotConfirmed',
  // Input validation
  'invalidDisplayName',
  'invalidName',
  'invalidTitle',
  'invalidDetails',
  'invalidInput',
  // Groups / membership
  'invalidCode',
  'rateLimited',
  'lastAdmin',
  'notMember',
  'cannotRemoveSelf',
  // Tasks
  'assigneeNotMember',
  'tooManyAssignees',
  // v2 (docs/CONTRACTS-V2.md §3)
  'invalidAppearance',
  'invalidRecurrence',
  'recurrenceNeedsDueDate',
  'invalidRotation',
  'invalidChecklistItem',
  'tooManyChecklistItems',
  // Generic
  'forbidden',
  'forbiddenFields',
  'notFound',
  'conflict',
  'network',
  'misconfigured',
  'unknown',
] as const;
export type AppErrorKind = (typeof APP_ERROR_KINDS)[number];

/**
 * The French message of an error, addressing the user with « tu » (docs/CONTRACTS-V2.md §13), with the French
 * typography of the app (’ and no-break spaces).
 */
export function messageFR(kind: AppErrorKind, detail: string | null = null): string {
  switch (kind) {
    case 'notAuthenticated':
      return 'Ta session a expiré. Reconnecte-toi.';
    case 'invalidCredentials':
      return 'E-mail ou mot de passe incorrect.';
    case 'emailAlreadyUsed':
      return 'Un compte existe déjà avec cet e-mail.';
    case 'weakPassword':
      return 'Mot de passe trop faible (8 caractères minimum).';
    case 'invalidEmail':
      return 'Adresse e-mail invalide.';
    case 'otpInvalid':
      return 'Code invalide ou expiré.';
    case 'emailRateLimited':
      return 'Trop d’e-mails envoyés. Réessaie plus tard.';
    case 'emailNotConfirmed':
      return 'Confirme d’abord ton adresse e-mail.';
    case 'invalidDisplayName':
      return 'Le nom doit contenir entre 1 et 50 caractères.';
    case 'invalidName':
      return 'Le nom du groupe doit contenir entre 1 et 60 caractères.';
    case 'invalidTitle':
      return 'Le titre doit contenir entre 1 et 200 caractères.';
    case 'invalidDetails':
      return 'La description ne doit pas dépasser 5000 caractères.';
    case 'invalidInput':
      return 'Certaines informations sont invalides.';
    case 'invalidCode':
      return 'Code d’invitation invalide.';
    case 'rateLimited':
      return 'Trop de tentatives. Réessaie dans une heure.';
    case 'lastAdmin':
      return 'Tu es l’unique admin\u{a0}: nomme d’abord un autre admin.';
    case 'notMember':
      return 'Cette personne ne fait pas partie du groupe.';
    case 'cannotRemoveSelf':
      return 'Pour te retirer du groupe, utilise «\u{a0}Quitter le groupe\u{a0}».';
    case 'assigneeNotMember':
      return 'Une personne assignée ne fait pas partie du groupe.';
    case 'tooManyAssignees':
      return '20 personnes assignées au maximum.';
    case 'invalidAppearance':
      return 'Couleur ou emoji invalide.';
    case 'invalidRecurrence':
      return 'Répétition invalide.';
    case 'recurrenceNeedsDueDate':
      return 'Choisis une échéance pour répéter la tâche.';
    case 'invalidRotation':
      return 'Le tour de rôle demande de 2 à 20 membres du groupe.';
    case 'invalidChecklistItem':
      return 'Un élément doit contenir entre 1 et 200 caractères.';
    case 'tooManyChecklistItems':
      return '30 éléments au maximum.';
    case 'forbidden':
      return 'Action non autorisée.';
    case 'forbiddenFields':
      return 'Tu peux seulement changer le statut de cette tâche.';
    case 'notFound':
      return 'Élément introuvable. Il a peut-être été supprimé.';
    case 'conflict':
      return 'Cet élément existe déjà.';
    case 'network':
      return 'Connexion impossible. Vérifie ton réseau.';
    case 'misconfigured':
      return 'L’application n’est pas configurée (Supabase).';
    case 'unknown':
      return `Une erreur est survenue. (${detail ?? ''})`;
  }
}

const BRAND = '__equipeAppError';

/** An error of the app: a kind, plus a French detail for `unknown`. Thrown by the core and the data layer. */
export class AppError extends Error {
  readonly kind: AppErrorKind;
  /** The detail of `unknown` (French, shown in parentheses); null for the other kinds. */
  readonly detail: string | null;
  readonly [BRAND] = true;

  constructor(kind: AppErrorKind, detail: string | null = null) {
    super(messageFR(kind, detail));
    this.name = 'AppError';
    this.kind = kind;
    this.detail = kind === 'unknown' ? (detail ?? '') : null;
  }

  static unknown(detail: string): AppError {
    return new AppError('unknown', detail);
  }

  /** The message shown to the user. */
  get messageFR(): string {
    return messageFR(this.kind, this.detail);
  }

  /** Same kind and same detail. */
  is(kind: AppErrorKind, detail?: string): boolean {
    return this.kind === kind && (detail === undefined || this.detail === detail);
  }

  equals(other: AppError): boolean {
    return this.kind === other.kind && this.detail === other.detail;
  }

  /** True for an `AppError`, even across bundles or transpiled classes. */
  static isAppError(value: unknown): value is AppError {
    return typeof value === 'object' && value !== null && (value as Record<string, unknown>)[BRAND] === true;
  }

  /** Wraps any error into an `AppError` (already-mapped errors pass through; a cancellation is « annulé »). */
  static wrap(error: unknown): AppError {
    if (AppError.isAppError(error)) return error;
    if (isCancellation(error)) return AppError.unknown('annulé');
    if (error instanceof Error) return AppError.unknown(error.message);
    return AppError.unknown(String(error));
  }
}

/** A cancelled request (an aborted fetch): never shown to the user. */
export function isCancellation(error: unknown): boolean {
  if (typeof error !== 'object' || error === null) return false;
  const name = (error as { name?: unknown }).name;
  return name === 'AbortError' || name === 'CancelledError';
}

/** The message to show for any error, or null for a cancellation (a cancelled load must not surface « annulé »). */
export function errorMessage(error: unknown): string | null {
  if (isCancellation(error)) return null;
  return AppError.wrap(error).messageFR;
}
